contracts/staking/ServiceStakingToken.sol

```diff
@@ -0,0 +1,99 @@
+    address public immutable stakingToken;
+
+    /// @dev ServiceStakingToken constructor.
+    /// @param _apy Staking APY (in single digits).
+    /// @param _minStakingDeposit Minimum security deposit for a service to be eligible to stake.
+    /// @param _stakingRatio Staking ratio: number of seconds per nonce (in 18 digits).
+    /// @param _serviceRegistry ServiceRegistry contract address.
+    /// @param _serviceRegistryTokenUtility ServiceRegistryTokenUtility contract address.
+    /// @param _stakingToken Address of a service security token.
+    constructor(
+        uint256 _apy,
+        uint256 _minStakingDeposit,
+        uint256 _stakingRatio,
+        address _serviceRegistry,
+        address _serviceRegistryTokenUtility,
+        address _stakingToken
+    )
+        ServiceStakingBase(_apy, _minStakingDeposit, _stakingRatio, _serviceRegistry)
+    {
+        // TODO: calculate minBalance
+        // Initial checks
+        if (_stakingToken == address(0) || _serviceRegistryTokenUtility == address(0)) {
+            revert ZeroAddress();
+        }
+
+        stakingToken = _stakingToken;
+        serviceRegistryTokenUtility = _serviceRegistryTokenUtility;
+    }
+
+    /// @dev Checks token security deposit.
+    /// @param serviceId Service Id.
+    function _checkTokenSecurityDeposit(uint256 serviceId) internal view override {
+        // Get the service security token and deposit
+        (address token, uint96 stakingDeposit) =
+            IServiceTokenUtility(serviceRegistryTokenUtility).mapServiceIdTokenDeposit(serviceId);
+
+        // The security token must match the contract token
+        if (stakingToken != token) {
+            revert();
+        }
+
+        // The security deposit must be greater or equal to the minimum defined one
+        if (stakingDeposit < minStakingDeposit) {
+            revert();
+        }
+    }
+
+    /// @dev Withdraws the reward amount to a service owner.
+    /// @param to Address to.
+    /// @param amount Amount to withdraw.
+    function _withdraw(address to, uint256 amount) internal override {
+        // Update the contract balance
+        balance -= amount;
+
+        SafeTransferLib.safeTransfer(stakingToken, to, amount);
+
+        emit Withdraw(to, amount);
+    }
+
+    /// @dev Deposits funds for staking.
+    /// @param amount Token amount to deposit.
+    function deposit(uint256 amount) external {
+        // Distribute current staking rewards
+        _checkpoint(0);
+
+        // Add to the overall balance
+        SafeTransferLib.safeTransferFrom(stakingToken, msg.sender, address(this), amount);
+
+        // Add to the contract and available rewards balances
+        uint256 newBalance = balance + amount;
+        uint256 newAvailableRewards = availableRewards + amount;
+
+        // Update rewards per second
+        uint256 newRewardsPerSecond = (newAvailableRewards * apy) / (100 * 365 days);
```

Comment:
I'd prefer if rewards per second was an arg rather than apy

Reply:
So as I understand what we want:
A given constant RewardsPerSecond (RPS) with dimension [T/s] T - token (including ETH)
RewardsPerSecond : immutable, and accordingly does not depend on the current AvailableRewards
Formula reward =  RPS * (t2 - t1) (1*)

A simple example of how this works without boundary conditions ( AvailableRewards >> sum(rewards)):
0. Start staking with AvailableRewards = 1M Tokens, no staking services, RPS = 1T/s
1. Staking 3 services: s1, s2, s3
2. +24h. Unstaking 3 services
s1 = pass tps_ratio => reward1 (yes)
s2 = pass tps_ratio => reward2 (yes)
s3 = fail tps_ratio => 0 (no)
reward1 = RPS * 24*60*60s = 1T/s * 86400s = 86400 T
reward2 = RPS * 24*60*60s = 1T/s * 86400s = 86400 T
AvailableRewards =  AvailableRewards - sum(rewards) = 1M - 2*86400 = 827,200

Let's assume we still have  AvailableRewards = 100k
0. No staking atm
1. Staking 3 services: s1, s2, s3
2. +24h. Unstaking 3 services
s1 = pass tps_ratio => reward1 (yes)
s2 = pass tps_ratio => reward2 (yes)
s3 = fail tps_ratio => 0 (no)
reward1 = RPS * 24*60*60s = 1T/s * 86400s = 86400 T (incorrect)
reward2 = RPS * 24*60*60s = 1T/s * 86400s = 86400 T (incorrect)
AvailableRewards =  AvailableRewards - sum(rewards) = 100k - 2*86400 < 0 
If AvailableRewards < sum(rewards) by (1*):
We need to distribute rewards based on the current limit in proportion to the "ideal" reward (1*)
reward1fix = RPS * 24*60*60s /sum(rewards) * AvailableRewards  = 1/2 * 100k = 50k T
reward2fix = RPS * 24*60*60s /sum(rewards) * AvailableRewards  = 1/2 * 100k = 50k T
AvailableRewards = AvailableRewards - 100k = 0

Reply:
With your suggestion @DavidMinarsch and the formula of @77ph the contract becomes significantly simpler. However, I have concerns because staking rewards are entirely unrelated to the service security deposit amount. This could potentially lead to a situation where staking rewards greatly exceed the service security deposits.

Reply:
"I have concerns because staking rewards are entirely unrelated to the service security deposit amount." -> that's up to the contract deployer. They set both after all.